File: core/ellipse.py
import cv2 as cv
import numpy as np


def ellipse_radius(ellipse):
    (_, _), (major, minor), _ = ellipse

    a = major / 2
    b = minor / 2

    return float(
        np.sqrt(a * b)
    )


def detect_outer_ellipse(gray):
    blur = cv.GaussianBlur(
        gray,
        (7, 7),
        0
    )

    _, thresh = cv.threshold(
        blur,
        0,
        255,
        cv.THRESH_BINARY_INV 
        + cv.THRESH_OTSU
    )

    kernel = cv.getStructuringElement(
        cv.MORPH_ELLIPSE,
        (9, 9)
    )

    thresh = cv.morphologyEx(
        thresh,
        cv.MORPH_CLOSE,
        kernel
    )

    contours, _ = cv.findContours(
        thresh,
        cv.RETR_EXTERNAL,
        cv.CHAIN_APPROX_SIMPLE
    )

    largest = max(
        contours,
        key=cv.contourArea
    )

    ellipse = cv.fitEllipse(
        largest
    )
    
    (
        (cx, cy),
        (major, minor),
        angle
    ) = ellipse

    if minor > major:
        major, minor = (
            minor,
            major
        )

        angle += 90
    
    ellipse = (
        (cx, cy),
        (
            major,
            minor
        ),
        angle
    )

    major_radius = major / 2
    minor_radius = minor / 2

    return {
        "ellipse": ellipse,
        "mask": thresh,

        "center": (
            float(cx),
            float(cy)
        ),

        "major_radius_px":
            float(
                major_radius
            ),

        "minor_radius_px":
            float(
                minor_radius
            ),

        "angle":
            float(angle),

        "radius_px":
            ellipse_radius(
                ellipse
            )
}

def detect_label_ellipse(
    image,
    spindle,
    rough_radius
):
    crop_radius = int(
        rough_radius * 0.45
    )

    x1 = max(
        spindle["x"]
        - crop_radius,
        0
    )

    y1 = max(
        spindle["y"]
        - crop_radius,
        0
    )

    x2 = min(
        spindle["x"]
        + crop_radius,
        image.shape[1]
    )

    y2 = min(
        spindle["y"]
        + crop_radius,
        image.shape[0]
    )

    crop = image[
        y1:y2,
        x1:x2
    ]

    hsv = cv.cvtColor(
        crop,
        cv.COLOR_BGR2HSV
    )

    lower = np.array(
        [5, 40, 40]
    )

    upper = np.array(
        [35, 255, 255]
    )

    mask = cv.inRange(
        hsv,
        lower,
        upper
    )

    contours, _ = cv.findContours(
        mask,
        cv.RETR_EXTERNAL,
        cv.CHAIN_APPROX_SIMPLE
    )

    largest = max(
        contours,
        key=cv.contourArea
    )

    ellipse = cv.fitEllipse(
        largest
    )

    (
        (cx, cy),
        (major, minor),
        angle
    ) = ellipse

    corrected = (
        (
            cx + x1,
            cy + y1
        ),
        (
            major,
            minor
        ),
        angle
    )

    return {
        "ellipse":
            corrected,
        "mask":
            mask,
        "radius_px":
            ellipse_radius(
                corrected
            )
    }

File: wasm/src/pipeline/ellipse.ts
// Mirrors backend/core/ellipse.py → detect_outer_ellipse()
// Input: grayscale Mat
// Output: VinylGeometry (center, radii, angle)
// All intermediate Mats deleted internally.

import type { VinylGeometry } from '../types';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function detectOuterEllipse(cv: any, gray: any): VinylGeometry {
  const blurred = new cv.Mat();
  try {
    cv.GaussianBlur(gray, blurred, new cv.Size(7, 7), 0, 0);

    const thresh = new cv.Mat();
    try {
      cv.threshold(blurred, thresh, 0, 255, cv.THRESH_BINARY_INV + cv.THRESH_OTSU);

      const kernel = cv.getStructuringElement(cv.MORPH_ELLIPSE, new cv.Size(9, 9));
      try {
        cv.morphologyEx(thresh, thresh, cv.MORPH_CLOSE, kernel);
      } finally {
        kernel.delete();
      }

      const contours = new cv.MatVector();
      const hierarchy = new cv.Mat();
      try {
        cv.findContours(thresh, contours, hierarchy, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE);

        const imageArea = gray.rows * gray.cols;
        let largestContour: any = null;
        let largestArea = 0;

        for (let i = 0; i < contours.size(); i++) {
          const c = contours.get(i);
          if (c.rows < 5) { c.delete(); continue; }
          const area = cv.contourArea(c);
          if (area < imageArea * 0.05) { c.delete(); continue; }
          if (area > largestArea) {
            if (largestContour) largestContour.delete();
            largestContour = c;
            largestArea = area;
          } else {
            c.delete();
          }
        }

        if (!largestContour) {
          throw new Error('No valid outer contour found — disc not detected');
        }

        try {
          const rrect = cv.fitEllipse(largestContour);
          // RotatedRect: { center: {x,y}, size: {width,height}, angle }
          let cx = rrect.center.x;
          let cy = rrect.center.y;
          let major = rrect.size.width;
          let minor = rrect.size.height;
          let angle = rrect.angle;

          // Normalize: ensure major >= minor (mirrors backend swap)
          if (minor > major) {
            [major, minor] = [minor, major];
            angle += 90;
          }

          const majorRadius = major / 2;
          const minorRadius = minor / 2;
          const radiusPx = Math.sqrt(majorRadius * minorRadius);

          return {
            center: [cx, cy],
            radiusPx,
            majorRadiusPx: majorRadius,
            minorRadiusPx: minorRadius,
            angle,
          };
        } finally {
          largestContour.delete();
        }
      } finally {
        hierarchy.delete();
        contours.delete();
      }
    } finally {
      thresh.delete();
    }
  } finally {
    blurred.delete();
  }
}
